### PR TITLE
Maybe make warrior lunge better[DNM]

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

Small refactor. Could make warrior lunge more reliable, or could do nothing.

closes https://github.com/tgstation/TerraGov-Marine-Corps/issues/6717 with some luck 

## Changelog
:cl:
fix: Warrior lunge more reliable
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm [hunk 1]
```diff
@@ -97,19 +99,24 @@
...
 
 	X.throw_at(get_step_towards(A, X), 6, 2, X)
-
-	if(distance > 1) //No zero timers
-		addtimer(CALLBACK(src, .proc/lunge_grab, X, A), min(0.05 SECONDS * max(0.01, distance - 1), 1 SECONDS)) //Minor delay so we're guaranteed to grab
-	else
-		lunge_grab(X,A)
+	lunge_target = A
+	RegisterSignal(X, COMSIG_MOVABLE_MOVED, .proc/check_if_lunge_possible)
+	check_if_lunge_possible(X, lunge_target)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:106]:
```suggestion
	check_if_lunge_possible(X)
```

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm [hunk 2]
```diff
@@ -60,6 +60,8 @@
 	cooldown_timer = 20 SECONDS
 	keybind_signal = COMSIG_XENOABILITY_LUNGE
 	target_flags = XABB_MOB_TARGET
+	/// The target of our lunge, we keep it to check if we are adjacent everytime we move
+	var/atom/lunge_target
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:64]: Don't see this being annulled

File: code/modules/mob/mob.dm [hunk 3]
```diff
@@ -447,14 +447,14 @@
 	return
 
 
-/mob/living/start_pulling(atom/movable/AM, suppress_message = FALSE)
+/mob/living/start_pulling(atom/movable/AM, suppress_message = FALSE, lunge = FALSE)
```
Comment (TiviPlus) [inline code/modules/mob/mob.dm:450]: Snowflake find a better way to do this

**Comment by SplinterGP** [comment]
im pretty sure the slowdown is due to how neck grab works.

**Review by TiviPlus — APPROVED** [review]
tested?

**Comment by BraveMole** [comment]
> tested?

yes i can lunge properly.
No clue if that fixes the bug since it's random

**Comment by Tupinambis** [comment]
Is this the bug where you lunge but don't actually land the lunge?

If so glad to see a fix.

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm [hunk 4]
```diff
@@ -97,24 +101,36 @@
...
+/datum/action/xeno_action/activable/lunge/proc/clean_lunge_target()
+	SIGNAL_HANDLER
+	deltimer(timer_id)
+	UnregisterSignal(lunge_target, COMSIG_PARENT_QDELETING)
+	lunge_target = null
+	UnregisterSignal(owner, COMSIG_MOVABLE_MOVED)
+	owner.remove_filter("warrior_lunge")
+	owner.stop_throw()
 
+/datum/action/xeno_action/activable/lunge/proc/lunge_grab(mob/living/carbon/xenomorph/warrior/X, atom/A)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:132]: autodoc

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm [hunk 5]
```diff
@@ -97,24 +101,36 @@
...
-	if(distance > 1) //No zero timers
-		addtimer(CALLBACK(src, .proc/lunge_grab, X, A), min(0.05 SECONDS * max(0.01, distance - 1), 1 SECONDS)) //Minor delay so we're guaranteed to grab
-	else
-		lunge_grab(X,A)
-
+	lunge_target = A
+	RegisterSignal(lunge_target, COMSIG_PARENT_QDELETING, .proc/clean_lunge_target)
+	RegisterSignal(X, COMSIG_MOVABLE_MOVED, .proc/check_if_lunge_possible)
+	check_if_lunge_possible(X)
+	timer_id = addtimer(CALLBACK(src, .proc/clean_lunge_target), 1 SECONDS, TIMER_STOPPABLE) //If we were blocked by a wall or a xeno while lunging, we clean everything
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:110]: what is this timer

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**